cpp/src/arrow/util/macros.h

```diff
@@ -49,10 +49,12 @@
 #define ARROW_PREDICT_TRUE(x) (__builtin_expect(!!(x), 1))
 #define ARROW_NORETURN __attribute__((noreturn))
 #define ARROW_NOINLINE __attribute__((noinline))
+#define ARROW_FORCE_INLINE __attribute__((always_inline))
 #define ARROW_PREFETCH(addr) __builtin_prefetch(addr)
 #elif defined(_MSC_VER)
 #define ARROW_NORETURN __declspec(noreturn)
 #define ARROW_NOINLINE __declspec(noinline)
+#define ARROW_FORCE_INLINE __declspec(forceinline)
```

Comment:
Also define an empty fallback below?

Reply:
Done. Interestingly `ARROW_NOINLINE` was missing the empty fallback.